source/common/network/lc_trie.h

```diff
@@ -0,0 +1,588 @@
+            buildRecursive(output.prefix_ + output.branch_, new_position, 1, address + i,
+                           next_free_index);
+          }
+          // Update the bit_pattern to skip over the trie_ entries initialized above.
+          bit_pattern += (1 << bits) - 1;
+        } else {
+          // Recursively build sub-tries for ip_prefixes_[new_position] to
+          // ip_prefixes_[new_position+count-1].
+          buildRecursive(output.prefix_ + output.branch_, new_position, count,
+                         address + bit_pattern, next_free_index);
+        }
+        new_position += count;
+      }
+    }
+
+    /**
+     * LcNode is a uint32_t. A wrapper is provided to simplify getting/setting the branch, the skip
+     * and the address values held within the structure.
+     *
+     * The LcNode has three parts to it
+     * - Branch: the first 5 bits represent the branching factor. The branching factor is used to
+     * determine the number of descendants for the current node. The number represents a power of 2,
+     * so there can be at most 2^31 descendant nodes.
+     * - Skip: the next 7 bits represent the number of bits to skip when looking at an IP address.
+     * This value can be between 0 and 127, so IPv6 is supported.
+     * - Address: the remaining 20 bits represent an index either into the trie_ or the
+     * ip_prefixes_. If branch_ != 0, the index is for the trie_. If branch == zero, the index is
+     * for the ip_prefixes_.
+     *
+     * Note: If more than 2^19-1 CIDR ranges are to be stored in trie_, uint64_t should be used
+     * instead.
+     */
+    struct LcNode {
+      uint32_t branch_ : 5;
+      uint32_t skip_ : 7;
+      uint32_t address_ : 20;
+    };
+
+    // Refer to LcNode to for further explanation on the current limitations for the maximum number
+    // of CIDR ranges supported and the maximum amount of nodes of supported in the trie.
+    static constexpr uint32_t MAXIMUM_CIDR_RANGE_ENTRIES = (1 << 19);
+
+    // During build(), an estimate of the number of nodes required will be made and set this value.
+    // This is used to ensure no out_of_range exception is thrown.
+    uint32_t maximum_trie_node_size;
+
+    // The CIDR range and tags needs to be maintained separately from the LC-Trie. A LC-Trie skips
+    // chunks of data while searching for a match. This means that the node found in the LC-Trie is
+    // not guaranteed to have the IP address in range. The last step prior to returning a tag is to
+    // check the CIDR range pointed to by the node in the LC-Trie has the IP address in range.
+    std::vector<IpPrefix<IpType>> ip_prefixes_;
+
+    // Main trie search structure.
+    std::vector<LcNode> trie_;
+
+    const double fill_factor_;
+    const uint32_t root_branching_factor_;
+  };
+
+  /**
+   * Converts Ipv6 address in std::array<uint8_t, 16> to absl::uint128 in host byte order.
+   * TODO (ccaraman): Remove this workaround once Ipv6Helper returns absl::uint128t.
```

Comment:
Can we just do that now? It doesn't seem that hard and seems like it would clean up a bunch of stuff. You could do another PR and just merge that into this?

Reply:
I'll do this in a separate PR and merge it into here.